libsoftwarecontainer/include/gateway.h

```diff
@@ -107,30 +93,51 @@ class Gateway: protected softwarecontainer::JSONParser
     virtual bool teardown();
 
     /**
-     * @brief Check if the gateway has an associated container
+     * @brief Set the associated container for this gateway
      */
-    bool hasContainer();
+    void setContainer(std::shared_ptr<ContainerAbstractInterface> container);
 
     /**
-     * @brief Get a reference to the associated container
+     * @brief Is the gateway configured or not?
      */
-    std::shared_ptr<ContainerAbstractInterface> getContainer();
+    bool isConfigured();
 
     /**
-     * @brief Set the associated contaitner for this gateway
+     * @brief Is the gateway activated or not?
      */
-    void setContainer(std::shared_ptr<ContainerAbstractInterface> container);
+    bool isActivated();
 
+protected:
     /**
-     * @brief Is the gateway configured or not?
+     * @brief Gateway specific parsing of config elements
+     *
+     * All gateways implement this method in order to provide gateway
+     * specific parsing of the configuration content.
+     *
+     * @param element A JSON configuration item.
+     * @returns ReturnCode FAILURE if an error was encountered while parsing, SUCCESS otherwise.
      */
-    bool isConfigured();
+    virtual ReturnCode readConfigElement(const json_t *element) = 0;
 
     /**
-     * @brief Is the gateway activated or not?
+     * @brief Check if the gateway has an associated container
+     *
+     * Inheriting gateways calls this method to know if they are safe to assume they
+     * can access a ContainerAbstractInterface instance previously set by SoftwareContainer
+     * calling setContainer()
+     *
+     * @returns true if there is a ContainerAbstractInterface instance set, false if not.
      */
-    bool isActivated();
+    bool hasContainer();
 
+    /**
+     * @brief Get a reference to the associated container
```

Comment:
I get why you do not write "shared pointer" but "reference" is a reserved word with a very specific meaning in C++ perhaps a better word can be found?